main.go

```diff
@@ -16,5 +16,155 @@
+	"github.com/networkservicemesh/sdk/pkg/tools/signalctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/spiffejwt"
+)
+
+// Config holds configuration parameters from environment variables
+type Config struct {
+	Name             string        `default:"icmp-server" desc:"Name of ICMP Server"`
+	BaseDir          string        `default:"./" desc:"base directory" split_words:"true"`
+	ListenOn         url.URL       `default:"unix:///listen.on.socket" desc:"url to listen on" split_words:"true"`
+	ConnectTo        url.URL       `default:"unix:///connect.to.socket" desc:"url to connect to" split_words:"true"`
+	MaxTokenLifetime time.Duration `default:"24h" desc:"maximum lifetime of tokens" split_words:"true"`
+
+	EndpointName string `default:"icmp-responder-nse" desc:"url to the local registry to register this NSE"`
+	CidrPrefix   string `default:"169.254.0.0/16" desc:"CIDR Prefix to assign IPs from"`
+}
+
 func main() {
+	// setup context to catch signals
+	ctx := signalctx.WithSignals(context.Background())
+	ctx, cancel := context.WithCancel(ctx)
+
+	// setup logging
+	logrus.SetFormatter(&nested.Formatter{})
+	logrus.SetLevel(logrus.TraceLevel)
+	ctx = log.WithField(ctx, "cmd", os.Args[0])
+
+	if err := debug.Self(); err != nil {
+		log.Entry(ctx).Infof("%s", err)
+	}
+
+	starttime := time.Now()
+
+	// get config from environment
+	config := &Config{}
+	if err := envconfig.Usage("nse", config); err != nil {
+		logrus.Fatal(err)
+	}
+	if err := envconfig.Process("nse", config); err != nil {
+		logrus.Fatalf("error processing config from env: %+v", err)
+	}
+
+	log.Entry(ctx).Infof("Config: %#v", config)
+
+	// get x509 source
+	source, err := workloadapi.NewX509Source(ctx)
+	if err != nil {
+		logrus.Fatalf("error getting x509 source: %+v", err)
+	}
+	svid, err := source.GetX509SVID()
+	if err != nil {
+		logrus.Fatalf("error getting x509 svid: %+v", err)
+	}
+	logrus.Infof("SVID: %q", svid.ID)
+
+	_, ipnet, err := net.ParseCIDR(config.CidrPrefix)
+	if err != nil {
+		log.Entry(ctx).Fatalf("error parsing cidr: %+v", err)
+	}
+
+	prefixes := []*net.IPNet{
+		ipnet,
+	}
+
+	ipamServer := point2pointipam.NewServer(prefixes...)
+
+	// icmp-server network service endpoint
+	responderEndpoint := endpoint.NewServer(
+		ctx,
+		config.Name,
+		authorize.NewServer(),
+		spiffejwt.TokenGeneratorFunc(source, config.MaxTokenLifetime),
+		ipamServer,
+		kernel.NewServer())
+
+	// create grpc server
+	// TODO add serveroptions for tracing
+	server := grpc.NewServer(grpc.Creds(credentials.NewTLS(tlsconfig.MTLSServerConfig(source, source, tlsconfig.AuthorizeAny()))))
+	responderEndpoint.Register(server)
+	srvErrCh := grpcutils.ListenAndServe(ctx, &config.ListenOn, server)
+	exitOnErr(ctx, cancel, srvErrCh)
+
+	var nsmTarget string
+	switch scheme := config.ConnectTo.Scheme; scheme {
+	case "tcp":
+		nsmTarget = config.ConnectTo.Host
+	default:
+		nsmTarget = config.ConnectTo.String()
+	}
+
+	cc, err := grpc.DialContext(ctx,
```

Comment:
You may want to call out here that you are registering the networkservice in this section.